Make the needs_reflow method actually work correctly.

eefriedman opened the pull request:
The document node is always dirty because layout never clears the bit;
instead, check the dirty bit of the root element.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/8441)

<!-- Reviewable:end -->

Ms2ger commented:
r? @bholley

bholley commented:
Wouldn't it be better to change the stuff in layout_task to operate on the Document node rather than the Document node's (presumably-only) first-child? Otherwise it seems like we'd have these weird asymmetries whereby we're constantly marking those bits on the document and never either inspecting or clearing them.

It was never clear to me why layout refers to the HTMLHTMLElement as the 'root node', since it would seem like it's the root element, and the Document is the root node. So a priori fixing that seems preferable, but there may be other considerations I'm not aware of.
<!-- Reviewable comment -K2mjbhHuk3Kw34LPGnU:-489068655 -->

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8441)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Maybe... and there's also the edge case of what to do with a document which doesn't have a root node.  On the other hand, there's a fundamental asymmetry: a Document doesn't really have any layout data associated with it.  I'll look a bit more.

eefriedman commented:
I spent a bit more time on this; the only solution I can think of which isn't completely confusing is associating a flow with the Document node, and that's kind of complicated...

If you want that solution, I'll just close this for now?

bholley commented:
Sorry for the terrible lag here. :-(

I think that associating a trivial flow with Document nodes makes sense (we now have LayoutDocument in the wrapper by which we can interact with Document, which should make everything easier). Unless @pcwalton has reason to believe it would be suboptimal for some reason.

So a patch to do that would be great. In the mean time though, this PR fixes a real correctness problem, so it's an improvement over the status quo. Let's get it landed.

bholley commented on components/script/dom/window.rs:
```diff
@@ -969,7 +969,8 @@ impl Window {
 
-        self.force_reflow(goal, query_type, reason)
+        self.force_reflow(goal, query_type, reason);
+        assert!(!self.Document().needs_reflow() || self.window_size.get().is_none());
```
Can you add a comment indicating why we need to weaken the assert case in the window_size.get().is_none() case?

bholley commented:
r+ with that fixed.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8569) made this pull request unmergeable. Please resolve the merge conflicts.

eefriedman committed:
Make the needs_reflow method actually work correctly.

The document node is always dirty because layout never clears the bit;
instead, check the dirty bit of the root element.

eefriedman commented:
@bors-servo r=bholley

bors-servo commented:
:pushpin: Commit 31c0138 has been approved by `bholley`

<!-- @bors-servo r=bholley 31c013858ff25e14f6984d4caa4d696d6aea33f8 -->

bors-servo commented:
:hourglass: Testing commit 31c0138 with merge a2be343...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/565), [gonk](http://build.servo.org/builders/gonk/builds/579), [linux-dev](http://build.servo.org/builders/linux-dev/builds/565), [linux-rel](http://build.servo.org/builders/linux-rel/builds/624), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/582), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/571), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/593)
